openapi3filter/validate_request.go

```diff
@@ -97,58 +98,85 @@ func ValidateParameter(c context.Context, input *RequestValidationInput, paramet
 	return nil
 }
 
+// ValidateRequestBody validates data of a request's body.
+// The function validates JSON data only.
+// The function returns nil when data is not JSON data.
+// The function returns RequestError with ErrInvalidRequired cause when a value is required but not defined.
+// The function returns RequestError with a openapi3.SchemaError cause when a value is invalid by JSON schema.
 func ValidateRequestBody(c context.Context, input *RequestValidationInput, requestBody *openapi3.RequestBody) error {
-	req := input.Request
-	content := requestBody.Content
-	if content != nil && len(content) > 0 {
-		inputMIME := req.Header.Get("Content-Type")
-		mediaType := parseMediaType(inputMIME)
-		contentType := requestBody.Content[mediaType]
-		if contentType == nil {
+	var (
+		req  = input.Request
+		data []byte
+	)
+
+	if req.Body != http.NoBody {
+		defer req.Body.Close()
+		var err error
+		if data, err = ioutil.ReadAll(req.Body); err != nil {
 			return &RequestError{
 				Input:       input,
 				RequestBody: requestBody,
-				Reason:      fmt.Sprintf("header 'Content-Type' has unexpected value: %q", inputMIME),
+				Reason:      "reading failed",
+				Err:         err,
 			}
 		}
-		schemaRef := contentType.Schema
-		if schemaRef != nil && isMediaTypeJSON(mediaType) {
-			schema := schemaRef.Value
-			body := req.Body
-			defer body.Close()
-			data, err := ioutil.ReadAll(body)
-			if err != nil {
-				return &RequestError{
-					Input:       input,
-					RequestBody: requestBody,
-					Reason:      "reading failed",
-					Err:         err,
-				}
-			}
+		// Put the data back into the input
+		req.Body = ioutil.NopCloser(bytes.NewReader(data))
+	}
 
-			// Put the data back into the input
-			req.Body = ioutil.NopCloser(bytes.NewReader(data))
+	if len(data) == 0 {
+		if requestBody.Required {
+			return &RequestError{Input: input, RequestBody: requestBody, Err: ErrInvalidRequired}
+		}
+		return nil
+	}
 
-			// Decode JSON
-			var value interface{}
-			if err := json.Unmarshal(data, &value); err != nil {
-				return &RequestError{
-					Input:       input,
-					RequestBody: requestBody,
-					Reason:      "decoding JSON failed",
-					Err:         err,
-				}
-			}
+	content := requestBody.Content
+	if len(content) == 0 {
+		// A request's body does not have declared content, so skip validation.
+		return nil
+	}
 
-			// Validate JSON with the schema
-			if err := schema.VisitJSON(value); err != nil {
-				return &RequestError{
-					Input:       input,
-					RequestBody: requestBody,
-					Reason:      "doesn't input the schema",
-					Err:         err,
-				}
-			}
+	inputMIME := req.Header.Get("Content-Type")
+	mediaType := parseMediaType(inputMIME)
+	if !isMediaTypeJSON(mediaType) {
+		// We support validation of JSON data only.
+		return nil
```

Comment:
I'm really not sure about this. People might want to define a "JSON Schema" (openapi3 JSON schema) to validate non-JSON payloads.
Also the OpenAPIv3 spec is starting to support alternative formats for schemas other than the bastardized version of JSON Schema Dratf-04 that's in use today: protocol buffers, XML stuff, and anything really.

So I'm thinking this should be an externally pluggable thing. A public map of MediaType to schema validator func? What do you think?

Reply:
Also on that map idea: if there's a map fetch miss then we should be returning something else than `nil` here.

Reply:
The minimum we can do about this for now is to return a non nil error here and later implement the public map mechanism or using context or whatever.

Reply:
Do you mean different formats for schemas or different formats for body's data? I think that support of different data' format will be useful.

I suggest to support content type `application/json` by default, and add function `openapi3filter.RegisterBodyDecoder`. Using this function users can register content types which is needed for them. For example:

```go
func main() {
	// ...

	// Register a body's decoder for content type "application/xml".
	openapi3filter.RegisterBodyDecoder("application/xml", xmlBodyDecoder)

	// Now HTTP request can contain a body with content type "application/xml".
	input := &openapi3filter.RequestValidationInput{
		Request:    httpReq,
		PathParams: pathParams,
		Route:      route,
	}
	if err := openapi3filter.ValidateRequest(ctx, input); err != nil {
		panic(err)
	}
}

func xmlBodyDecoder(body []byte) (interface{}, error) {
	// Decode body to a primitive, []inteface{}, or map[string]interface{}.
}
```

Internally the library will use registered decoders to decode a body of HTTP request and response. After that the decoded value will be validated by JSON Schema.

Reply:
I like this RegisterBodyDecoder thing. I think we still need an additional RegisterBodyValidator too but that may be left as a todo for the future.

Reply:
Done.